### PR TITLE
[SYCL] Support query of free device memory extension

### PR DESCRIPTION
**Pull request by smaslov-intel**
Support new extensions to query free device memory with `info::device::ext_intel_device_info_free_memory`
E2E test https://github.com/intel/llvm-test-suite/pull/1162
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/doc/extensions/supported/sycl_ext_intel_device_info.md [hunk 1]
```diff
@@ -347,3 +348,41 @@ Then the maximum memory bandwidth can be obtained using the standard get\_info()
...
+## Version ##
+
+The extension supports this query in version 4 and later.
+
+
+## Device Information Descriptors ##
+
+| Device Descriptors | Return Type | Description |
+| ------------------ | ----------- | ----------- |
+| info\:\:device\:\:ext\_intel\_device\_info\_free\_memory | uint64\_t| Returns the memory avialble on the device in units of bytes.|
```
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:369]: This name seems redundant because "device" and "info" are repeated.  Currently, the name of the descriptor is like:

```
sycl::info::device::ext_intel_device_info_free_memory
```

It seems like we could shorten this to:

```
sycl::info::device::ext_intel_free_memory
```

Actually, you should also coordinate with Rana Rauf, who is changing the namespace of these info descriptors.  After that change the descriptor would be like:

```
sycl::ext::intel::info::device::free_memory
```

(Essentially, we use the `sycl::ext::intel` namespace instead of the `ext_intel_` prefix.)

I'm not sure if it's better for you to add this descriptor with the old name, and then let Rana change it to the new style vs. you adding this new descriptor now with the new style.  Either way is OK with me as long as it ends up with the new style.

Comment (smaslov-intel) [reply sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:369]: I thought about this too, but then the the `dev.has(aspect::ext_intel_free_memory)` could become ambiguous since it wouldn't have `device` and `info` in it. Today we have a mixture of everything and I am not adding more mess. Can change to anything that we define as a standard.

**Comment by smaslov-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1162

File: sycl/doc/extensions/supported/sycl_ext_intel_device_info.md (hunk 1 above)
Comment (gmlueck) [reply sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:369]: It looks like all of the existing queries in this spec have a consistent naming except for the "uuid" query.  Those queries omit both "info" and "device" from the name of the descriptor and from the name of the aspect.  Following that style, you would have:

```
if (dev.has(aspect::ext_intel_free_memory)) {
  auto FreeMemory = dev.get_info<ext::intel::info::device::free_memory>();
}
```

I agree that there could be some confusion about the aspect name.  We could add the word "info" to the aspect (only) to avoid this confusion:

```
if (dev.has(aspect::ext_intel_info_free_memory)) {
  auto FreeMemory = dev.get_info<ext::intel::info::device::free_memory>();
}
```

There is no confusion about "device" because all aspects refer to a device.

If we like that naming style, we can change the other aspect names to match, leaving the existing ones as deprecated.  We could also update the "uuid" query to follow this style.

Comment (gmlueck) [reply sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:369]: I don't feel that strongly, so let's match the style of the majority of the existing queries and use this style:

```
if (dev.has(aspect::ext_intel_free_memory)) {
  auto FreeMemory = dev.get_info<ext::intel::info::device::free_memory>();
}
```

File: sycl/include/sycl/detail/pi.h [hunk 2]
```diff
@@ -271,6 +273,7 @@ typedef enum {
   PI_DEVICE_INFO_IMAGE_SRGB = 0x10027,
   // Return true if sub-device should do its own program build
   PI_DEVICE_INFO_BUILD_ON_SUBDEVICE = 0x10028,
+  PI_EXT_INTEL_DEVICE_INFO_FREE_MEMORY = 0x10029,
```
Comment (steffenlarsen) [inline sycl/include/sycl/detail/pi.h:276]: Other EXT enumerations step the value up, i.e. `PI_EXT_ONEAPI_DEVICE_INFO_*` starts at `0x20000`. I think it would be best to do that with `PI_EXT_INTEL_DEVICE_INFO_*` too. Either as a collective `PI_EXT_*_DEVICE_INFO_*` numbering, or by taking another leap in numbers, for example `0x30000` for `PI_EXT_INTEL_DEVICE_INFO_*`. I think I prefer the latter.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -3106,6 +3106,27 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
                   ZeDevicePciProperties.address.function);
     return ReturnValue(AddressBuffer);
   }
+
+  case PI_EXT_INTEL_DEVICE_INFO_FREE_MEMORY: {
+    if (getenv("ZES_ENABLE_SYSMAN") == nullptr) {
+      zePrint("Set ZES_ENABLE_SYSMAN=1 to obtain free memory.\n");
+      return PI_ERROR_INVALID_VALUE;
```
Comment (steffenlarsen) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3113]: Should this be a backend specific error instead, so the user gets notified rather than it looking like a backend crash?

File: sycl/doc/extensions/supported/sycl_ext_intel_device_info.md [hunk 4]
```diff
@@ -347,3 +348,41 @@ Then the maximum memory bandwidth can be obtained using the standard get\_info()
     if (dev.has(aspect::ext_intel_max_mem_bandwidth)) {
       auto maxBW = dev.get_info<info::device::ext_intel_max_mem_bandwidth>();
     }
+
+# Free Memory #
+
+A new device descriptor will be added which will provide the number of bytes of free memory for the device.
```
Comment (steffenlarsen) [inline sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:354]: Maybe a little pedantic, but should this specify that it is global memory?

File: sycl/include/sycl/detail/pi.h (hunk 2 above)
Comment (smaslov-intel) [reply sycl/include/sycl/detail/pi.h:276]: Other PI*_DEVICE_INFO* around where I put new one, are already Intel extension (since line 263), they just weren't accurate to use PI_EXT_INTEL as a prefix. Not sure if finding another unused range is meaningful then

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3113]: I can do that, so that the message is shown as a part of SYCL trace, not just with ZE_DEBUG=1 currently. Although it is rather a warning, not an error. Basically, in case the ZES_ENABLE_SYSMAN is not set to 1 the has_aspect, which users must check, would return false.

File: sycl/doc/extensions/supported/sycl_ext_intel_device_info.md (hunk 4 above)
Comment (smaslov-intel) [reply sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:354]: OK

File: sycl/include/sycl/detail/pi.h (hunk 2 above)
Comment (steffenlarsen) [reply sycl/include/sycl/detail/pi.h:276]: I see! In that case the current value seems fitting. Thanks for clarifying.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Added it as a plugin-specific message

**Comment by smaslov-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1162

**Review by steffenlarsen — DISMISSED** [review]
LGTM!

**Comment by smaslov-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1162

**Comment by smaslov-intel** [comment]
The CUDA failure is unrelated:

> Starting job container
  /usr/bin/docker --config /localdisk2/github/_work/_temp/.docker_b91ba181-41a3-4ff7-b2f2-fdd2c4da1958 login ghcr.io -u smaslov-intel --password-stdin
  Error: Docker login for 'ghcr.io' failed with exit code 1

@steffenlarsen : Please re-approve/merge this.

**Review by steffenlarsen — APPROVED** [review]

**Comment by steffenlarsen** [comment]
Approval still needed from @intel/dpcpp-l0-pi-reviewers .

**Comment by smaslov-intel** [comment]
> Approval still needed from @intel/dpcpp-l0-pi-reviewers .

@againull : please review the small change in the L0 plugin

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -3129,6 +3129,28 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
...
+    ZE_CALL(zesDeviceEnumMemoryModules, (ZeDevice, &MemCount, nullptr));
+    std::vector<zes_mem_handle_t> MemHandles(MemCount);
+    ZE_CALL(zesDeviceEnumMemoryModules,
+            (ZeDevice, &MemCount, MemHandles.data()));
+
+    for (auto &ZesMemHandle : MemHandles) {
+      ZesStruct<zes_mem_state_t> ZeMemState;
+      ZE_CALL(zesMemoryGetState, (ZesMemHandle, &ZeMemState));
+      FreeMemory += ZeMemState.free;
+    }
```
Comment (againull) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3150]: Should we take into account only specific type of memory here?
For example, the list of memory types includes ZES_MEM_TYPE_SLM and ZES_MEM_TYPE_GRF: https://spec.oneapi.io/level-zero/latest/sysman/api.html?highlight=zesdeviceenummemorymodules#memory-enums
In other words, is it correct to treat all types of memory as  the global device memory in this case?

**Review by steffenlarsen — APPROVED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I've added a check HBM type as that is where zeMemAllocDevice allocates from.

Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Thank you!

**Review by againull — APPROVED** [review]

**Comment by zjin-lcf** [comment]
Thank you for the extension. I have two questions.

Could you please explain why "intel" is included in the "info::device::ext_intel_device_info_free_memory" ?
Could the function return both free memory and total memory ?
